rero_ils/modules/documents/dojson/contrib/jsontomarc21/model.py

```diff
@@ -62,15 +61,142 @@ def set_value(data, old_data, key):
 
 
+def replace_concept_sources(concept, source_order):
+    """Prepare concept data from sources.
+
+    :param concept: concept to use.
+    :source_order: Source order to use to get the informations.
+    :returns: concept entity with localized values.
+    """
+    def set_value(data, old_data, key):
+        """Set new data value."""
+        value = data.get(key)
+        if not old_data.get(key) and value:
+            old_data[key] = value
+        return old_data
+
+    refs = []
+    for source in source_order:
+        source_data = concept.get(source, {})
+        if source_data:
+            refs.append({
+                'source': source,
+                'pid': source_data.get("pid")
+            })
+            for key in ['type', 'authorized_access_point']:
+                concept = set_value(source_data, concept, key)
+            concept.pop(source)
+    concept['refs'] = refs
+    return concept
+
+
+def do_contribution(contribution, source_order):
+    """Create contribution.
+
+    :param contribution: contribution to transform.
+    :param source_order: source_order to use.
+    :returns: result marc dictionary,
+              entity type,
+              is it a surname,
+              is it a conference
+    """
+    roles = contribution.get('role', [])
+    entity = contribution.get('entity')
+    if pid := entity.get('pid'):
+        # we have a $ref, get the real entity
+        ref = entity.get('$ref')
+        if entity := Entity.get_record_by_pid(pid):
+            contribution = replace_contribution_sources(
+                contribution={'entity': entity},
+                source_order=source_order
+            )
+            entity = contribution['entity']
+        else:
+            current_app.logger.error(
+                f'No entity found for pid:{pid} {ref}')
+            return None, None, False, False
+    preferred_name = entity.get('preferred_name')
+    if not preferred_name:
+        current_app.logger.warning(
+            f'JSON to MARC21 contribution no preferred_name: {entity}')
+        preferred_name = entity.get(
+            f'authorized_access_point_{to_marc21.language}')
+    result = {}
+    conference = False
+    surname = False
+    result = add_value(result, 'a', preferred_name)
+    entity_type = entity.get('type')
+    if entity_type == EntityType.PERSON:
+        if ',' in preferred_name:
+            surname = True
+        result = add_value(result, 'b', entity.get('numeration'))
+        result = add_value(result, 'c', entity.get('qualifier'))
+        date_of_birth = entity.get('date_of_birth', '')
+        date_of_birth = \
+            date_of_birth[:4] if len(date_of_birth) > 3 else ''
+        date_of_death = entity.get('date_of_death', '')
+        date_of_death = \
+            date_of_death[:4] if len(date_of_death) > 3 else ''
+        date = f'{date_of_birth} - {date_of_death}'
+        if date != ' - ':
+            result = add_value(result, 'd', date)
```

Comment:
```python
date_parts = [
    entity['date_of_birth'][:4] if len(entity.get('date_of_birth', '')) > 3 else None,
    entity['date_of_death'][:4] if len(entity.get('date_of_death', '')) > 3 else None
]
if date_parts := filter(None, date_parts):
    result = add_value(result, 'd', ' - '.join(date_parts))
```

Reply:
This is unfortunately not working if only the death date is given.

Reply:
Ok, So it's even simplier :
```python
dates = ' - '.join([
    entity['date_of_birth'][:4] if len(entity.get('date_of_birth', '')) > 3 else '',
    entity['date_of_death'][:4] if len(entity.get('date_of_death', '')) > 3 else ''
])
result = add_value(result, 'd', dates)
```